### PR TITLE
feat: Integrate Atoma with Phi3

### PR DESCRIPTION
**Pull request by fishonamos**

### CODE REVIEW

File: src/models/mod.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
 pub mod llama;
+pub mod phy3;
 
 pub use llama::Llama;
+pub use phy3::Model as Phi3;
```
Comment (jorgeantonio21) [inline src/models/mod.rs:5]: It's not `phy` but `phi`

File: src/lib.rs [hunk 2]
```diff
@@ -6,3 +6,4 @@ pub use flash_attention::{
     FlashAttentionPrefillMetadata,
 };
 pub use models::Llama;
+pub use models::phy3::Model as Phi3;
```
Comment (jorgeantonio21) [inline src/lib.rs:9]: * `Phi`

File: src/models/phy3.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+#[derive(Debug, Clone)]
+struct Attention {
+    qkv_proj: Linear,
+    o_proj: Linear,
+    num_heads: usize,
+    num_kv_heads: usize,
+    num_kv_groups: usize,
+    head_dim: usize,
+    rotary_emb: Arc<RotaryEmbedding>,
+    kv_cache: Option<(Tensor, Tensor)>,
```
Comment (jorgeantonio21) [inline src/models/phy3.rs:81]: We do not use `kv_cache` here, this is kept within the GPU own memory (which is allocated on fresh start). Please take a look at the Llama implementation.

File: src/models/phy3.rs [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+        let (key_states, value_states) = match &self.kv_cache {
+            None => (key_states, value_states),
+            Some((prev_k, prev_v)) => {
+                let key_states = Tensor::cat(&[prev_k, &key_states], 2)?;
+                let value_states = Tensor::cat(&[prev_v, &value_states], 2)?;
+                (key_states, value_states)
+            }
+        };
+        self.kv_cache = Some((key_states.clone(), value_states.clone()));
```
Comment (jorgeantonio21) [inline src/models/phy3.rs:157]: This is not valid in our setup, see comments above.

File: src/models/phy3.rs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+        let attn_output = self.attention.forward(&query_states, &key_states, &value_states, &self.kv_cache.as_ref().unwrap().0, attention_metadata)?;
+
+        attn_output
+            .transpose(1, 2)?
+            .reshape((b_sz, q_len, ()))?
+            .apply(&self.o_proj)
+    }
+
+    fn clear_kv_cache(&mut self) {
```
Comment (jorgeantonio21) [inline src/models/phy3.rs:171]: Not applicable in our case, see comments above

File: src/models/phy3.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            embed_tokens,
+            layers,
+            norm,
+            lm_head,
+            device: vb.device().clone(),
+            dtype: vb.dtype(),
+        })
+    }
+
+    pub fn forward(&mut self, input_ids: &Tensor, seqlen_offset: usize, attention_metadata: &FlashAttentionMetadata) -> Result<Tensor> {
```
Comment (jorgeantonio21) [inline src/models/phy3.rs:291]: Don't we need `input_positions` to apply rotary embeddings ?

File: src/models/phy3.rs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+            .apply(&self.lm_head)
+    }
+
+    pub fn clear_kv_cache(&mut self) {
+        for layer in self.layers.iter_mut() {
+            layer.clear_kv_cache()
+        }
+    }
+
+    fn prepare_decoder_attention_mask(
```
Comment (jorgeantonio21) [inline src/models/phy3.rs:314]: I don't think we need decoding mask, this should be handled by the flash attention kernels, if used properly.

**Review by jorgeantonio21 — COMMENTED** [review]
It is going in the right direction, but it needs to be reworked to be compatible with the current usage of flash attention kernels and paged attention.

**Comment by fishonamos** [comment]
Made the updates as reviewed. @jorgeantonio21.  Removed kv cache, decoder mask and applied input positions to rotary embeddings.

**Review by jorgeantonio21 — APPROVED** [review]
LGTM
